### DISCUSSION TITLE
[Detector Support]: Hailo-8 support missing in HAOS add-on

### DISCUSSION THREAD

**Discussion by sairon** [blakeblackshear/frigate#15172] [Detector Support]
### Describe the problem you are having

Frigate's main Docker image doesn't include Hailo's userspace drivers which are needed for Hailo-8(L) support. This means its not possible to use the accelerator if Frigate is deployed as an HAOS add-on which uses this image (FTR it needs the full access version to access the hardware).

Another issue is that HAOS doesn't provide a way to build custom drivers, so it depends on what is included in the upstream Raspberry Pi kernel, which is [currently 4.18.0](https://github.com/raspberrypi/linux/tree/stable_20241008/drivers/media/pci/hailo). The driver and userspace version (and the firmware as well) must match, otherwise initialization fails with an error.

For easier usage, aarch64 image might track what's currently available in Raspberry Pi's current stable kernel (which is tracked by HAOS as well) so the image would work "out of the box" without the need to compile custom driver version, etc. Alternatively, it should be possible to customize the userspace version and download it in runtime instead of during the Docker build. This will also remove the need for having separate images as the Hailo runtime binaries and Python dependencies are the only difference from the standard image.

### Version

0.15.0-beta2

### Frigate config file

```yaml
...

detectors:
  hailo8l:
    type: hailo8l
    device: PCIe
    model:
      path: /config/model_cache/h8l_cache/ssd_mobilenet_v1.hef

...
```

### docker-compose file or Docker CLI command

```yaml
n/a
```

### Relevant Frigate log output

```shell
2024-11-25 10:31:51.372063344  Process detector:hailo8l:
2024-11-25 10:31:51.372065492  Traceback (most recent call last):
2024-11-25 10:31:51.372066659    File "/opt/frigate/frigate/detectors/plugins/hailo8l.py", line 72, in __init__
2024-11-25 10:31:51.372067399      self.target = VDevice()
2024-11-25 10:31:51.372068233  NameError: name 'VDevice' is not defined
2024-11-25 10:31:51.372068825  
2024-11-25 10:31:51.372070103  During handling of the above exception, another exception occurred:
2024-11-25 10:31:51.372071973  
2024-11-25 10:31:51.372073047  Traceback (most recent call last):
2024-11-25 10:31:51.372096955    File "/usr/lib/python3.9/multiprocessing/process.py", line 315, in _bootstrap
2024-11-25 10:31:51.372098029      self.run()
2024-11-25 10:31:51.372099121    File "/opt/frigate/frigate/util/process.py", line 41, in run_wrapper
2024-11-25 10:31:51.372099955      return run(*args, **kwargs)
2024-11-25 10:31:51.372100992    File "/usr/lib/python3.9/multiprocessing/process.py", line 108, in run
2024-11-25 10:31:51.372122899      self._target(*self._args, **self._kwargs)
2024-11-25 10:31:51.372124121    File "/opt/frigate/frigate/object_detection.py", line 121, in run_detector
2024-11-25 10:31:51.372136918      object_detector = LocalObjectDetector(detector_config=detector_config)
2024-11-25 10:31:51.372138047    File "/opt/frigate/frigate/object_detection.py", line 68, in __init__
2024-11-25 10:31:51.372139066      self.detect_api = create_detector(detector_config)
2024-11-25 10:31:51.372140288    File "/opt/frigate/frigate/detectors/__init__.py", line 18, in create_detector
2024-11-25 10:31:51.372141103      return api(detector_config)
2024-11-25 10:31:51.372142214    File "/opt/frigate/frigate/detectors/plugins/hailo8l.py", line 105, in __init__
2024-11-25 10:31:51.372143010      except HailoRTException as e:
2024-11-25 10:31:51.372154158  NameError: name 'HailoRTException' is not defined
```

### Install method

HassOS Addon

### Object Detector

Other

### Screenshots of the Frigate UI's System metrics pages

n/a

**Comment by NickM-27** [comment]
For now this is working as expected, HA OS including support was a very recent change.

I also don't think we want to be downloading these at runtime, anytime a user updates the image or clears the docker container it will have to be downloaded again as it will only be stored in the docker layer which is something that is not recommended and that we've been sure to avoid. 

Frigate also supports running hailo on amd64 and we wouldn't want these to be treated differently, I could see us moving the version back so it matches the RPi kernel so that a specific Hailo addon could be added, but would need to talk to @blakeblackshear about that.

**Reply by sairon** [reply]
> For now this is working as expected, HA OS including support was a very recent change.

Right, that's on us :see_no_evil:

> I also don't think we want to be downloading these at runtime, anytime a user updates the image or clears the docker container it will have to be downloaded again as it will only be stored in the docker layer which is something that is not recommended and that we've been sure to avoid.

Actually that was an idea that occurred to me when I was writing the post, I didn't think that through completely. Similar thing is done to the model but in that case it is downloaded to a bind-mounted volume (config folder), so it's not entirely the same.

> Frigate also supports running hailo on amd64 and we wouldn't want these to be treated differently, I could see us moving the version back so it matches the RPi kernel so that a specific Hailo addon could be added, but would need to talk to @blakeblackshear about that.

In fact it's what [I did](https://github.com/sairon/frigate/commit/1f7dc4d5eedc2a41722f55b499858a334f2b74ef) to test it. I just feel it's a bit unfortunate everything will be reliant on Raspberry Pi kernel release cycle (which can be sometimes quite slow). Also it usually takes a while before the kernel makes its way to HAOS, so when the version gets updated in Frigate, users will be locked to use particular HAOS version (or version of the `linux` package in the RPi OS), and vice versa.

Moreover, if it were possible to install the libraries on-demand, it will be another argument for dropping the dedicated `-h8l` image.

Also, on a slightly related note, to test everything I had to fork the [hailort repo](https://github.com/sairon/hailort) as well and rebase it to build the older revision. There I had the idea it would be maybe more manageable if the repo wasn't a fork with the extra build tooling added, but a standalone repo just with the build scripts and the source `hailort` repository as a submodule that can be changed to any arbitrary revision as needed. But that's just my 2 cents.

**Reply by NickM-27** [reply]
For now I don't really see an advantage to having it all built in to the same docker image, especially when the full access addon is required (at least currently). I think if the hailo device becomes more popular than we can always reconsider including the dependencies in the default image, but for now having a dedicated addon image should work without issues assuming we get the driver / FW versions synced up.

**Reply by sairon** [reply]
Okay, I'm leaving it up to you, anything that makes it working as an add-on works for me.

With my HAOS maintainer hat on, I can say that we currently track what Raspberry Pi provides in their sources and stable releases. But again, it should be noted that it always takes couple of weeks before Raspberry Pi's kernel version lands in HAOS, as we prefer to wait for some general testing to be done by the RPi OS users rather than turning our users into the testing cohort. As more than 40 % of HAOS deployments (from those 400k that opted-in for the analytics) are running on RPi, merging the Linux kernel, which always means several upstream Linux versions and thousands of commits of diff here, is a bit sensitive.

Eventually, if it turns out to be unbearable, we could consider building a custom version of the driver out-of-tree. In fact, it would also help us a bit, because currently we need to sync the [firmware version](https://github.com/home-assistant/operating-system/commit/eec6dfb28d575966802866b116050335d1b803fb) with the driver in the RPi's Linux kernel. And that way we could also add support for x86 HAOS targets (as they can easily use the M.2 modules).
